File: src/server/vterm/vterm.cc
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/log/Services.h>
#include <base/util/BitField.h>

#include <m3/com/MemGate.h>
#include <m3/server/Server.h>
#include <m3/server/RequestHandler.h>
#include <m3/session/ServerSession.h>
#include <m3/vfs/GenericFile.h>

using namespace m3;

static constexpr size_t MSG_SIZE    = 64;
static constexpr size_t BUF_SIZE    = 256;
static constexpr size_t MAX_CLIENTS = 32;

struct VTermSession;
class ChannelSession;
class VTermHandler;

using base_class = RequestHandler<
    VTermHandler, GenericFile::Operation, GenericFile::Operation::COUNT, VTermSession
>;

static Server<VTermHandler> *srv;

struct VTermSession : public ServerSession {
    enum Type {
        META,
        CHAN,
    };

    explicit VTermSession(capsel_t srv_sel, capsel_t caps)
        : ServerSession(srv_sel, caps) {
    }
    virtual ~VTermSession() {
    }

    virtual Type type() const = 0;

    virtual void next_in(m3::GateIStream &is) {
        reply_error(is, m3::Errors::NOT_SUP);
    }
    virtual void next_out(m3::GateIStream &is) {
        reply_error(is, m3::Errors::NOT_SUP);
    }
    virtual void commit(m3::GateIStream &is, size_t) {
        reply_error(is, m3::Errors::NOT_SUP);
    }
};

class MetaSession : public VTermSession {
public:
    explicit MetaSession(capsel_t srv_sel, capsel_t caps = ObjCap::INVALID)
        : VTermSession(srv_sel, caps) {
    }

    ChannelSession *create_chan(bool write);

    virtual Type type() const {
        return META;
    }
};

class ChannelSession : public VTermSession {
public:
    explicit ChannelSession(uint id, RecvGate &rgate, MemGate &mem,
                            capsel_t srv_sel, capsel_t caps, bool _writing)
        : VTermSession(srv_sel, caps),
          id(id),
          active(false),
          writing(_writing),
          ep(ObjCap::INVALID),
          sgate(SendGate::create(&rgate, reinterpret_cast<label_t>(this), MSG_SIZE, nullptr, caps + 1)),
          mem(mem.derive(id * BUF_SIZE, BUF_SIZE, MemGate::RW)),
          pos(),
          len() {
    }
    ~ChannelSession();

    ChannelSession *clone();

    virtual Type type() const override {
        return CHAN;
    }

    virtual void next_in(m3::GateIStream &is) override {
        SLOG(VTERM, "[" << id << "] vterm::next_in()");

        if(writing) {
            reply_error(is, Errors::NO_PERM);
            return;
        }

        pos += len - pos;

        Errors::last = Errors::NONE;
        if(pos == len) {
            char buf[BUF_SIZE];
            len = static_cast<size_t>(Machine::read(buf, sizeof(buf)));
            mem.write(buf, len, 0);
            pos = 0;
        }

        if(!active) {
            Syscalls::activate(ep, mem.sel(), 0);
            active = true;
        }

        if(Errors::last != Errors::NONE)
            reply_error(is, Errors::last);
        else
            reply_vmsg(is, Errors::NONE, pos, len - pos);
    }

    virtual void next_out(m3::GateIStream &is) override {
        SLOG(VTERM, "[" << id << "] vterm::next_out()");

        if(!writing) {
            reply_error(is, Errors::NO_PERM);
            return;
        }

        if(len > 0) {
            char buf[BUF_SIZE];
            mem.read(buf, len, 0);
            Machine::write(buf, len);
            len = 0;
        }

        if(!active) {
            Syscalls::activate(ep, mem.sel(), 0);
            active = true;
        }

        if(Errors::last != Errors::NONE)
            reply_error(is, Errors::last);
        else {
            pos = 0;
            len = BUF_SIZE;
            reply_vmsg(is, Errors::NONE, static_cast<size_t>(0), BUF_SIZE);
        }
    }

    virtual void commit(m3::GateIStream &is, size_t nbytes) override {
        SLOG(VTERM, "[" << id << "] vterm::commit(nbytes=" << nbytes << ")");

        if(nbytes > len - pos) {
            reply_error(is, Errors::INV_ARGS);
            return;
        }

        if(writing) {
            if(len > 0) {
                char buf[BUF_SIZE];
                mem.read(buf, nbytes, 0);
                Machine::write(buf, nbytes);
                len = 0;
            }
            reply_vmsg(is, Errors::NONE);
        }
        else {
            pos += nbytes;
            reply_vmsg(is, Errors::NONE);
        }
    }

    uint id;
    bool active;
    bool writing;
    capsel_t ep;
    SendGate sgate;
    MemGate mem;
    size_t pos;
    size_t len;
};

class VTermHandler : public base_class {
public:
    explicit VTermHandler(WorkLoop *wl)
        : base_class(),
          _slots(),
          _mem(MemGate::create_global(MAX_CLIENTS * BUF_SIZE, MemGate::RW)),
          _rgate(RecvGate::create(nextlog2<MAX_CLIENTS * MSG_SIZE>::val, nextlog2<MSG_SIZE>::val)) {
        add_operation(GenericFile::SEEK, &VTermHandler::invalid_op);
        add_operation(GenericFile::STAT, &VTermHandler::invalid_op);
        add_operation(GenericFile::NEXT_IN, &VTermHandler::next_in);
        add_operation(GenericFile::NEXT_OUT, &VTermHandler::next_out);
        add_operation(GenericFile::COMMIT, &VTermHandler::commit);
        add_operation(GenericFile::CLOSE, &VTermHandler::close_chan);

        using std::placeholders::_1;
        _rgate.start(wl, std::bind(&VTermHandler::handle_message, this, _1));
    }

    virtual Errors::Code open(VTermSession **sess, capsel_t srv_sel, word_t) override {
        *sess = new MetaSession(srv_sel);
        return Errors::NONE;
    }

    virtual Errors::Code obtain(VTermSession *sess, KIF::Service::ExchangeData &data) override {
        if(data.caps != 1 && data.caps != 2)
            return Errors::INV_ARGS;

        ChannelSession *nsess;
        if(sess->type() == VTermSession::META) {
            if(data.args.count != 1)
                return Errors::INV_ARGS;
            nsess = static_cast<MetaSession*>(sess)->create_chan(data.args.vals[0] == 1);
        }
        else {
            if(data.args.count != 0)
                return Errors::INV_ARGS;
            nsess = static_cast<ChannelSession*>(sess)->clone();
        }
        if(nsess == nullptr)
            return Errors::NO_SPACE;

        data.caps = KIF::CapRngDesc(KIF::CapRngDesc::OBJ, nsess->sel(), 2).value();
        return Errors::NONE;
    }

    virtual Errors::Code delegate(VTermSession *sess, KIF::Service::ExchangeData &data) override {
        if(data.caps != 1 || data.args.count != 0 || sess->type() != VTermSession::CHAN)
            return Errors::INV_ARGS;

        ChannelSession *chan = static_cast<ChannelSession*>(sess);
        chan->ep = VPE::self().alloc_sel();
        data.caps = KIF::CapRngDesc(KIF::CapRngDesc::OBJ, chan->ep, 1).value();
        return Errors::NONE;
    }

    virtual Errors::Code close(VTermSession *sess) override {
        delete sess;
        return Errors::NONE;
    }

    virtual void shutdown() override {
        _rgate.stop();
    }

    void invalid_op(GateIStream &is) {
        reply_vmsg(is, m3::Errors::NOT_SUP);
    }

    void next_in(m3::GateIStream &is) {
        VTermSession *sess = is.label<VTermSession*>();
        sess->next_in(is);
    }

    void next_out(m3::GateIStream &is) {
        VTermSession *sess = is.label<VTermSession*>();
        sess->next_out(is);
    }

    void commit(m3::GateIStream &is) {
        VTermSession *sess = is.label<VTermSession*>();
        size_t nbytes;
        is >> nbytes;

        sess->commit(is, nbytes);
    }

    void close_chan(m3::GateIStream &is) {
        VTermSession *sess = is.label<VTermSession*>();
        close(sess);
        reply_error(is, Errors::NONE);
    }

    ChannelSession *new_chan(bool write) {
        // find and allocate free slot
        uint id = _slots.first_clear();
        if(id == MAX_CLIENTS)
            return nullptr;
        _slots.set(id, true);

        capsel_t caps = VPE::self().alloc_sels(2);
        return new ChannelSession(id, _rgate, _mem, srv->sel(), caps, write);
    }

    BitField<MAX_CLIENTS> _slots;
    MemGate _mem;
    RecvGate _rgate;
};

inline ChannelSession *MetaSession::create_chan(bool write) {
    return srv->handler().new_chan(write);
}

inline ChannelSession::~ChannelSession() {
    srv->handler()._slots.set(id, false);
}
inline ChannelSession *ChannelSession::clone() {
    return srv->handler().new_chan(writing);
}

int main() {
    WorkLoop wl;

    srv = new Server<VTermHandler>("vterm", &wl, new VTermHandler(&wl));
    wl.run();
    delete srv;

    return 0;
}

File: src/server/vterm/src/vterm.rs
/*
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#![no_std]

mod chan;
mod input;

use m3::cap::SelSpace;
use m3::cell::LazyStaticRefCell;
use m3::col::Vec;
use m3::com::{opcodes, GateIStream, MemGate, Perm, RGateArgs, RecvGate};
use m3::errors::{Code, Error};
use m3::io::LogFlags;
use m3::kif;
use m3::log;
use m3::mem::GlobOff;
use m3::rc::Rc;
use m3::server::{
    server_loop, CapExchange, ClientManager, ExcType, RequestHandler, RequestSession, Server,
    ServerSession, SessId, DEF_MAX_CLIENTS,
};
use m3::tiles::Activity;

static MEM: LazyStaticRefCell<Rc<MemGate>> = LazyStaticRefCell::default();

#[allow(clippy::large_enum_variant)]
#[derive(Debug)]
enum SessionData {
    Meta,
    Chan(chan::Channel),
}

#[derive(Debug)]
pub struct VTermSession {
    _serv: ServerSession,
    data: SessionData,
    parent: Option<SessId>,
    childs: Vec<SessId>,
}

impl RequestSession for VTermSession {
    fn new(serv: ServerSession, _arg: &str) -> Result<Self, Error>
    where
        Self: Sized,
    {
        Ok(VTermSession {
            _serv: serv,
            data: SessionData::Meta,
            parent: None,
            childs: Vec::new(),
        })
    }

    fn close(&mut self, cli: &mut ClientManager<Self>, sid: SessId, sub_ids: &mut Vec<SessId>)
    where
        Self: Sized,
    {
        log!(
            LogFlags::VTReqs,
            "[{}] vterm::close(): closing {:?}",
            sid,
            sub_ids
        );

        // close child sessions as well
        sub_ids.extend_from_slice(&self.childs);

        // remove us from parent
        if let Some(pid) = self.parent.take() {
            if let Some(p) = cli.get_mut(pid) {
                p.childs.retain(|cid| *cid != sid);
            }
        }
    }
}

impl VTermSession {
    fn get_sess(cli: &mut ClientManager<Self>, sid: SessId) -> Result<&mut Self, Error> {
        cli.get_mut(sid).ok_or_else(|| Error::new(Code::InvArgs))
    }

    fn with_chan<F, R>(&mut self, is: &mut GateIStream<'_>, func: F) -> Result<R, Error>
    where
        F: Fn(&mut chan::Channel, &mut GateIStream<'_>) -> Result<R, Error>,
    {
        match &mut self.data {
            SessionData::Meta => Err(Error::new(Code::InvArgs)),
            SessionData::Chan(c) => func(c, is),
        }
    }

    fn new_chan(parent: SessId, serv: ServerSession, writing: bool) -> Result<VTermSession, Error> {
        log!(LogFlags::VTReqs, "[{}] vterm::new_chan()", serv.id());

        Ok(VTermSession {
            data: SessionData::Chan(chan::Channel::new(
                serv.id(),
                MEM.borrow().clone(),
                writing,
            )?),
            _serv: serv,
            parent: Some(parent),
            childs: Vec::new(),
        })
    }

    fn clone(
        cli: &mut ClientManager<Self>,
        crt: usize,
        sid: SessId,
        xchg: &mut CapExchange<'_>,
    ) -> Result<(), Error> {
        log!(LogFlags::VTReqs, "[{}] vterm::clone(crt={})", sid, crt);

        let (sel, _nsid) = cli.add_connected(crt, |cli, serv, _sgate| {
            let parent_sess = Self::get_sess(cli, sid)?;
            let nsid = serv.id();

            let child_sess = match &parent_sess.data {
                SessionData::Meta => {
                    let writing = xchg.in_args().pop::<i32>()? == 1;
                    Self::new_chan(sid, serv, writing)
                },

                SessionData::Chan(c) => Self::new_chan(sid, serv, c.is_writing()),
            }?;

            // remember that the new session is a child of the current one
            parent_sess.childs.push(nsid);
            Ok(child_sess)
        })?;

        xchg.out_caps(kif::CapRngDesc::new(kif::CapType::Object, sel, 2));
        Ok(())
    }

    fn set_dest(
        cli: &mut ClientManager<Self>,
        _crt: usize,
        sid: SessId,
        xchg: &mut CapExchange<'_>,
    ) -> Result<(), Error> {
        log!(LogFlags::VTReqs, "[{}] vterm::set_dest()", sid);

        let sess = Self::get_sess(cli, sid)?;
        match &mut sess.data {
            SessionData::Chan(c) => {
                let sel = SelSpace::get().alloc_sel();
                c.set_dest(sel);
                xchg.out_caps(kif::CapRngDesc::new(kif::CapType::Object, sel, 1));
                Ok(())
            },
            _ => Err(Error::new(Code::InvArgs)),
        }
    }

    fn enable_notify(
        cli: &mut ClientManager<Self>,
        _crt: usize,
        sid: SessId,
        xchg: &mut CapExchange<'_>,
    ) -> Result<(), Error> {
        log!(LogFlags::VTReqs, "[{}] vterm::set_notify()", sid);

        let sess = Self::get_sess(cli, sid)?;
        match &mut sess.data {
            SessionData::Chan(c) => {
                let rgate = RecvGate::new_with(RGateArgs::default().order(6).msg_order(6))?;
                let sgate = c.set_notify_gates(rgate)?;

                xchg.out_caps(kif::CapRngDesc::new(kif::CapType::Object, sgate, 1));
                Ok(())
            },
            _ => Err(Error::new(Code::InvArgs)),
        }
    }
}

#[no_mangle]
pub fn main() -> Result<(), Error> {
    MEM.set(Rc::new(
        MemGate::new((DEF_MAX_CLIENTS * chan::BUF_SIZE) as GlobOff, Perm::RW)
            .expect("Unable to alloc memory"),
    ));

    let mut hdl = RequestHandler::new().expect("Unable to create request handler");
    let srv = Server::new("vterm", &mut hdl).expect("Unable to create service 'vterm'");

    use opcodes::File;
    hdl.reg_cap_handler(File::CloneFile, ExcType::Obt(2), VTermSession::clone);
    hdl.reg_cap_handler(File::SetDest, ExcType::Del(1), VTermSession::set_dest);
    hdl.reg_cap_handler(
        File::EnableNotify,
        ExcType::Del(1),
        VTermSession::enable_notify,
    );

    hdl.reg_msg_handler(File::NextIn, |sess, is| {
        sess.with_chan(is, |c, is| c.next_in(is))
    });
    hdl.reg_msg_handler(File::NextOut, |sess, is| {
        sess.with_chan(is, |c, is| c.next_out(is))
    });
    hdl.reg_msg_handler(File::Commit, |sess, is| {
        sess.with_chan(is, |c, is| c.commit(is))
    });
    hdl.reg_msg_handler(File::FStat, |sess, is| {
        sess.with_chan(is, |c, is| c.stat(is))
    });
    hdl.reg_msg_handler(File::Seek, |_sess, _is| Err(Error::new(Code::NotSup)));
    hdl.reg_msg_handler(File::GetTMode, |sess, is| {
        sess.with_chan(is, |c, is| c.get_tmode(is))
    });
    hdl.reg_msg_handler(File::SetTMode, |sess, is| {
        sess.with_chan(is, |c, is| c.set_tmode(is))
    });
    hdl.reg_msg_handler(File::ReqNotify, |sess, is| {
        sess.with_chan(is, |c, is| c.request_notify(is))
    });

    let sel = SelSpace::get().alloc_sel();
    let serial_gate = Activity::own()
        .resmng()
        .unwrap()
        .get_serial(sel)
        .expect("Unable to allocate serial rgate");

    server_loop(|| {
        srv.fetch_and_handle(&mut hdl)?;

        if let Ok(msg) = serial_gate.fetch() {
            input::handle_input(hdl.clients_mut(), msg);
            serial_gate.ack_msg(msg).unwrap();
        }

        input::receive_acks(hdl.clients_mut());

        hdl.fetch_and_handle_msg();

        Ok(())
    })
    .ok();

    Ok(())
}
